addons/udes_permissions/models/user_template.py

```diff
@@ -0,0 +1,303 @@
+    # Be wary if changing this if we have existing data, as data will not migrate with it
+    TEMPLATE_GROUP_PREFIX = "User Template Group: "
+
+    # Actual columns the user can configure
+    name = fields.Char(
+        string="Template Name",
+        required=True,
+        help="""The name of the User Template,
+        typically something like Administrator, Stock Manager, etc..""",
+    )
+    complete_name = fields.Char("Complete Name", compute="_compute_complete_name", store=True)
+    group_ids = fields.Many2many(
+        "res.groups",
+        string="Groups",
+        required=True,
+        help="""The Groups for this User Template,
+        groups here, along with all child templates groups will
+        be considered in get_template_groups()""",
+    )
+    # NB: We don't cascade on parent as it won't call the unlink() hook where cascaded
+    parent_id = fields.Many2one(
+        "user.template",
+        index=True,
+        string="Parent",
+        help="""The Parent Template of this User Template.
+        Allows us to create a hierarchy of templates.
+        A User Template can have multiple child templates (forks in the tree)""",
+    )
+    # Magic column updated as _parent_store is set to True
+    parent_path = fields.Char(index=True)
+
+    # Inverse to parent_id
+    child_ids = fields.One2many(
+        "user.template",
+        "parent_id",
+        string="Children",
+        readonly=True,
+        help="""User Templates which have directly have this template set as their parent.
+        All templates groups in the family (excluding parents) will be considered
+        when the User Template is applied to a user (via the family_group_ids column)""",
+    )
+
+    # Magic column that gets written to by the system on create. cascade only removes the
+    # template if we delete the group, so we need an additional hook in unlink() on user.template
+    template_group_id = fields.Many2one(
+        "res.groups",
+        ondelete="cascade",
+        readonly=True,
+        required=True,
+        help="""A new group is created and attached to a User Template on template creation
+        which allows for configuring specific ACLs for a template without too much hassle,
+        with the downside of not being able to revoke specific ACLs from that group from a user.
+        This can be used as an interim until more granular groups are available.
+        The group set here will be bundled in with all other groups as part of get_template_groups()""",
+    )
+
+    # The three following computes all use the same method to save searching child_of
+    # on three separate occasions, as they all hinge off the same initial 'family'
+    family_descendant_ids = fields.Many2many(
+        "user.template",
+        compute="_compute_family_column_ids",
+        store=False,  # We cannot store as function won't recompute if indirect descendants groups change
```

Comment:
I don't mind being explicit, but `store=False` is the default. Is this to make sure no one tries this? :D

Reply:
yeah it's more to explicitly prevent anyone from ever trying it in future because it will not work :+1: